### PR TITLE
ARROW-3610: [C++] Add interface to turn stl_allocator into arrow::MemoryPool

### PR DESCRIPTION
**Pull request by xhochy**
First outcome from PyConDE collaboration with @wolfv

### CODE REVIEW

File: cpp/src/arrow/allocator.h [hunk 1]
```diff
@@ -85,6 +85,41 @@ class stl_allocator {
...
 };
 
+template <typename Allocator = std::allocator<uint8_t>>
+class STLMemoryPool : public MemoryPool {
+ public:
+  explicit STLMemoryPool(const Allocator& alloc) : alloc_(alloc) {}
+  ~STLMemoryPool() override = default;
+
+  Status Allocate(int64_t size, uint8_t** out) override {
+    *out = alloc_.allocate(size);
```
Comment (pitrou) [inline cpp/src/arrow/allocator.h]: Shouldn't we try to catch `std::bad_alloc` here and return `OutOfMemory`?

File: cpp/src/arrow/allocator.h [hunk 2]
```diff
@@ -85,6 +85,41 @@ class stl_allocator {
...
+    stats_.UpdateAllocatedBytes(size);
+    return Status::OK();
+  }
+
+  Status Reallocate(int64_t old_size, int64_t new_size, uint8_t** ptr) override {
+    uint8_t* old_ptr = *ptr;
+    *ptr = alloc_.allocate(new_size);
+    memcpy(*ptr, old_ptr, std::min(old_size, new_size));
+    Free(old_ptr, old_size);
```
Comment (pitrou) [inline cpp/src/arrow/allocator.h]: This will wrongly call `UpdateAllocatedBytes`.

File: cpp/src/arrow/allocator.h [hunk 3]
```diff
@@ -85,6 +85,41 @@ class stl_allocator {
...
+
+  Status Allocate(int64_t size, uint8_t** out) override {
+    *out = alloc_.allocate(size);
+    stats_.UpdateAllocatedBytes(size);
+    return Status::OK();
+  }
+
+  Status Reallocate(int64_t old_size, int64_t new_size, uint8_t** ptr) override {
+    uint8_t* old_ptr = *ptr;
+    *ptr = alloc_.allocate(new_size);
```
Comment (pitrou) [inline cpp/src/arrow/allocator.h]: Also catch `std::bad_alloc` here.

File: cpp/src/arrow/allocator-test.cc [hunk 4]
```diff
@@ -23,9 +23,26 @@
...
+#include "arrow/test-util.h"
 
 namespace arrow {
 
+TEST(STLMemoryPool, Base) {
+  std::allocator<uint8_t> allocator;
+  STLMemoryPool<std::allocator<uint8_t>> pool(allocator);
+
+  uint8_t* data = nullptr;
+  ASSERT_OK(pool.Allocate(100, &data));
```
Comment (pitrou) [inline cpp/src/arrow/allocator-test.cc]: You should also test `Reallocate` and its effect on `bytes_allocated`.

**Review by pitrou — CHANGES_REQUESTED** [review]
Looks good on the principle (TBH I am flabbergasted that STL allocators don't support resizing...).

File: cpp/src/arrow/allocator.h [hunk 5]
```diff
@@ -85,6 +85,41 @@ class stl_allocator {
   MemoryPool* pool_;
 };
 
+template <typename Allocator = std::allocator<uint8_t>>
+class STLMemoryPool : public MemoryPool {
+ public:
+  explicit STLMemoryPool(const Allocator& alloc) : alloc_(alloc) {}
+  ~STLMemoryPool() override = default;
```
Comment (wesm) [inline cpp/src/arrow/allocator.h]: This dtor may not be needed

File: cpp/src/arrow/memory_pool.h [hunk 6]
```diff
@@ -25,6 +25,35 @@
...
+    auto allocated = bytes_allocated_.fetch_add(diff) + diff;
+    // "maximum" allocated memory is ill-defined in multi-threaded code,
+    // so don't try to be too rigorous here
+    if (diff > 0 && allocated > max_memory_) {
+      max_memory_ = allocated;
+    }
+  }
+
+ protected:
+  std::atomic<int64_t> bytes_allocated_;
```
Comment (wesm) [inline cpp/src/arrow/memory_pool.h]: Need to include atomic in this header now

File: cpp/src/arrow/allocator.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/allocator.h]: Let's say that before reallocate, you have `bytes_allocated = old_size`.
After reallocate, you want to have `bytes_allocated = new_size`.
But in reality, you have:
* after the call to `Free(old_ptr, old_size)`: `bytes_allocated = 0`
* after the call to `UpdateAllocatedBytes(new_size - old_size)`: `bytes_allocated = new_size - old-size`.

And this is visible in the test results btw: https://travis-ci.org/apache/arrow/jobs/447042467#L1094

Comment (xhochy) [reply cpp/src/arrow/allocator.h]: 🤦‍♂️ I was running the release tests while having a debug build. That explains my confusion.

Comment (xhochy) [reply cpp/src/arrow/allocator.h]: Fixed now

**Comment by xhochy** [comment]
The test failure is because I have removed the `DCHECK` from the MemoryPoolStats while I moved them to the header. As we don't allow `DCHECK` in the header, I'm currently tending to removing this test.

**Comment by wesm** [comment]
I think it's OK to remove that test

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
The "death test" is failing on AppVeyor.

**Comment by xhochy** [comment]
AppVeyor partly succeeded here: https://ci.appveyor.com/project/xhochy/arrow/builds/20006144 

Seems like there is some AppVeyor-internal problem currently.

**Pull request closed without merging**